Document key type upgrades (#883)
digital-asset/docs.daml.com: docs/2.10.0/docs/concepts/ledger-model/upgrade-model-in-depth.rst

Template Keys
~~~~~~~~~~~~~

The key of the new version of a template must be a valid upgrade of the key of
the prior version of that template.

Adding a key, removing a key, or changing the type of the key for a non-valid

> upgrade leads to a validation error.

Reviewer: Can we express this without the use of "non-valid upgrade" (i.e. the negation)? I found that a quick read/parse of this statement quickly leads to misunderstandings.

Reviewer: OR maybe we bring "non-valid upgrade" to the sentence start to make it more prominent?

Author: I broke it into two sentences.

.. _examples-3:

**Examples**

Below, the template on the right is a valid upgrade of the template on the left:

> the type of the key of the template on the right is a valid upgrade of the type

Reviewer: ```suggestion
the type of the key of the template on the right is a valid upgrade of the type
```
```suggestion
this is because the type of the key of the template on the right is a valid upgrade of the type
```

Template Keys
~~~~~~~~~~~~~

The key of the new version of a template must be a valid upgrade of the key of
the prior version of that template.

Adding a key or removing a key leads to a validation error.

> Changing the type of a key for a type that is not a valid upgrade of the
> original type leads to a validation error.

Reviewer 2: ```suggestion
Changing the key type to one that is not a valid upgrade of the
original type leads to a validation error.
```

.. _examples-3:

**Examples**

> Below, the template on the right is a valid upgrade of the template on the left:
> this is because the type of the key of the template on the right is a valid upgrade of the type
> of the key of the template on the left.

Reviewer 2: ```suggestion
Below, the template on the right is a valid upgrade of the template on the left because the key type of the template on the right is a valid upgrade of the key type of the template on the left:
```

                signatory p
        
Below, the template on the right is **not** a valid upgrade of the

> template on the left because it changes the type of its key for a type that
> is not a valid upgrade of ``(Party, Text)``.

Reviewer 2: ```suggestion
template on the left because it changes the key type for a type that
is not a valid upgrade of ``(Party, Text)``.
```

- The contract key;
- The maintainers of the contract key.

The metadata of two contracts are equivalent if and only if:

> - Their contract IDs are equal;
> - Their signatories are equal;
> - Their stakeholders are equal;
> - Their keys, after transformation to the maximum version of the two contracts, are equal.
> - Their maintainers are equal;

Reviewer: ```suggestion
- their contract IDs are equal;
- their signatories are equal;
- their stakeholders are equal;
- their keys, after transformation to the maximum version of the two contracts, are equal;
- their maintainers are equal.
```

             where
               signatory sig, sig
    
Assume the same leger as above. Fetching contract ``1234`` with target type

> ``p-2.0.0:T`` retrieves the the contract and again successfully

Reviewer: ```suggestion
``p-2.0.0:T`` retrieves the contract and again successfully
```

transforms it into the value ``T { sig = 'Alice', additionalSig = None }``. The
signatories of this transformed contract are then computed using the expression
``sig, sig``, which evaluate to the list ``['Alice', 'Alice']``. This list is
then compared to signatories of the original contract stored on the ledger:
``['Alice']``. They do not match and thus the upgrade is rejected at runtime.

**Example 2**

> Below the module on the right is a valid upgrade of the module on the left.

Reviewer 2: ```suggestion
Below the module on the right is a valid upgrade of the module on the left:
```

               p : Party

           template T 
             with
               sig : Party
             where
               signatory sig
               key MyKey p : MyKey
               maintainer key.p

     - .. code-block:: daml
           
           module M where

           data MyKey = MyKey
             with
               p : Party
               i : Optional Int

           template T 
             with
               sig : Party
               i : Optional Int
             where
               signatory sig
               key MyKey p i : MyKey
               maintainer key.p
    
Assume a ledger that contains a contract of type ``T`` written by
``p-1.0.0``.

+-------------+---------------+---------------------------+-------------------------+
| Contract ID | Contract Type | Contract Key              | Contract                |
+=============+===============+===========================+=========================+
| ``1234``    | ``p-1.0.0:T`` | ``MyKey { p = 'Alice' }`` | ``T { sig = 'Alice' }`` |
+-------------+---------------+---------------------------+-------------------------+

Fetching contract ``1234`` with package preference ``p-2.0.0`` retrieves the

> contract and successfully transforms it into a value of type ``p-2.0.0:T``: ``T
> { sig = 'Alice', i = None }``. The key of this transformed contract is then

Reviewer: I found this confusing on first read. Perhaps:
```suggestion
contract and successfully transforms it into a value ``T
{ sig = 'Alice', i = None }`` of type ``p-2.0.0:T``. The key of this transformed contract is then
```

computed using the expression ``MyKey p i``, which evaluates to the value
``MyKey { p = 'Alice', i = None }``. This value is then compared against the key

> of the original contract after transformation to a value of type
> ``p-2.0.0:MyKey``: ``MyKey { p = 'Alice', i = None }``. The two values match and

Reviewer: Simarly, maybe:
```suggestion
of the original contract after transformation to a value ``MyKey { p = 'Alice', i = None }`` of type
``p-2.0.0:MyKey``. The two values match and
```

thus the upgrade is valid.

> On the other hand, below, the module on the right is **not** a valid upgrade

Reviewer: ```suggestion
On the other hand, see below, the module on the right is **not** a valid upgrade
```

> of the module on the left.

Reviewer 2: ```suggestion
of the module on the left:
```

.. list-table::
   :widths: 50 50
   :width: 100%
   :class: diff-block

   * -  In ``p-1.0.0``:
     -  In ``p-2.0.0``:

   * - .. code-block:: daml 
           
           module M where

           data MyKey = MyKey
             with
               p : Party

           template T 
             with
               sig : Party
             where
               signatory sig
               key MyKey p : MyKey
               maintainer key.p

     - .. code-block:: daml
           
           module M where

           data MyKey = MyKey
             with
               p : Party
               i : Optional Int

           template T 
             with
               sig : Party
             where
               signatory sig
               key MyKey p (Some 0) : MyKey
               maintainer key.p

> Assume the same leger as above. Fetching contract ``1234`` with package

Reviewer: ```suggestion
Assume the same ledger as above. Fetching contract ``1234`` with package
```

preference ``p-2.0.0`` retrieves the the contract and again successfully
transforms it into the value ``T { sig = 'Alice' }``. The key of this
transformed contract is then computed using the expression ``MyKey p (Some 0)``,

> which evaluates to the value ``MyKey { p = 'Alice', i = Some 0 }``. This value is
> then compared against the key of the original contract after transformation to a
> value of type ``p-2.0.0:MyKey``: ``MyKey { p = 'Alice', i = None }``. The two

Reviewer 2: ```suggestion
which evaluates to the value ``MyKey { p = 'Alice', i = Some 0 }``. This value is
then compared against the original contract's key after transformation to a
value of type ``p-2.0.0:MyKey``: ``MyKey { p = 'Alice', i = None }``. The two
```

> value of type ``p-2.0.0:MyKey``: ``MyKey { p = 'Alice', i = None }``. The two

Reviewer: ```suggestion
value ``MyKey { p = 'Alice', i = None }`` of type ``p-2.0.0:MyKey``. The two
```

> values don't match and thus the upgrade is rejected at runtime.

Reviewer 2: ```suggestion
values do not match and thus the upgrade is rejected at runtime.
```

into a value of type ``p-2.0.0:T``: ``T { sig = 'Alice', i = 42 }`` and its view
is computed again, this time according to ``p-2.0.0``: ``IView 43``. Because the
two views differ, the fetch is rejected at runtime.

Key Transformation in FetchByKey and LookupByKey
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

When fetching or looking up a contract by key in the body of a choice, the type
of the key expression is known at compile time. Let us call it the target type.
Then the returned contract, if any, verifies the property that its key, after

> transformation to the target type, matches the key used for querying.

Author: Done

> Then the returned contract, if any, verifies the property that its key, after
> transformation to the target type, matches the key used for querying.

Reviewer 2: ```suggestion
The returned contract, if any, verifies that its key, after
transformation to the target type, matches the key used for querying.
```

**Example**

> Below the module on the right is a valid upgrade of the module on the left.

Reviewer 2: ```suggestion
Below the module on the right is a valid upgrade of the module on the left:
```

             with
               sig : Party
               i : Optional Int
             where
               signatory sig
               key MyKey p i : MyKey
               maintainer key.p
    
Assume a ledger that contains a contract of type ``T`` written by
``p-1.0.0``.

+-------------+---------------+---------------------------+-------------------------+
| Contract ID | Contract Type | Contract Key              | Contract                |
+=============+===============+===========================+=========================+
| ``1234``    | ``p-1.0.0:T`` | ``MyKey { p = 'Alice' }`` | ``T { sig = 'Alice' }`` |
+-------------+---------------+---------------------------+-------------------------+

Finally, assume a third package which depends on ``p-2.0.0`` and defines a
choice involving the following lookup by key:

.. code:: daml

  choice C : ()
    controller ctl
    do
      cid <- lookupByKey @T (MyKey alice None)
      ...

The static type of the key being looked up is thus ``p-2.0.0:MyKey``. The key of
contract ``1234``, once transformed to a value of type ``p-2.0.0:MyKey``,
becomes ``MyKey { p = 'Alice', i = None }``. This matches the key being looked

> up, so ``cid`` will be bound to ``Some 1234``.

Reviewer 2: ```suggestion
up, so ``cid`` is bound to ``Some 1234``.
```
